### PR TITLE
Make PUT shutdown API idempotent

### PR DESCRIPTION
**Pull request by AthenaEryma**
This PR makes the PUT shutdown API idempotent, as well as allowing
switching from shutdown types.

### CODE REVIEW

File: x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportPutShutdownNodeAction.java [hunk 1]
```diff
@@ -135,6 +135,18 @@ public void onFailure(Exception e) {
         });
     }
 
+    private boolean isTypeChangeAllowed(SingleNodeShutdownMetadata existingRecord, PutShutdownNodeAction.Request request) {
```
Comment (dakrone) [inline x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportPutShutdownNodeAction.java:138]: We could probably make this static and unit test it, what do you think?

**Review by dakrone — APPROVED** [review]
LGTM, I left one comment

File: x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportPutShutdownNodeAction.java (hunk 1 above)
Comment (gwbrown) [reply x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportPutShutdownNodeAction.java]: Good call, I switched this to take 2 `Type`s (it doesn't need the whole metadata or request) and added some unit tests.

**Comment by gwbrown** [comment]
@dakrone You good with the changes in https://github.com/elastic/elasticsearch/pull/75161/commits/c5361efa2d7bc45847316c2700774f479dfa69a6 as well?

File: x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportPutShutdownNodeAction.java [hunk 2]
```diff
@@ -72,9 +70,16 @@ public ClusterState execute(ClusterState currentState) {
                 }
 
                 // Verify that there's not already a shutdown metadata for this node
-                if (Objects.nonNull(currentShutdownMetadata.getAllNodeMetadataMap().get(request.getNodeId()))) {
-                    logger.error(Strings.toString(currentShutdownMetadata));
-                    throw new IllegalArgumentException("node [" + request.getNodeId() + "] is already shutting down");
+                SingleNodeShutdownMetadata existingRecord = currentShutdownMetadata.getAllNodeMetadataMap().get(request.getNodeId());
+                if (existingRecord != null) {
+                    logger.trace(
+                        "replacing existing shutdown record for node [{}] of type [{}] with reason [{}] with new type [{}] and reason [{}]",
```
Comment (dakrone) [inline x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportPutShutdownNodeAction.java:76]: I think this is a rare enough occurrence, and useful enough, to update to this:

```suggestion
                    logger.info(
                        "updating existing shutdown record for node [{}] of type [{}] with reason [{}] with new type [{}] and reason [{}]",
```

**Review by dakrone — APPROVED** [review]
LGTM, I left one super minor comment

**Pull request merged**